### PR TITLE
feat: Add ability to get the native Element of a DOMElement

### PR DESCRIPTION
**Pull request by alexanderGugel**
* Add Compositor#stripNonCloneableOutCommands
* Add Compositor#message
* Add DOMElement#getEl

Non-cloneable objects are also being pushed onto the _outCommands
queue, but their index will be stored separately.
stripNonCloneableOutCommands is then being used in order to prepare the
command queue for postMessage.

@DnMllr I don't think there is a cleaner solution for this use case.

### CODE REVIEW

**Comment by michaelobriena** [comment]
@alexanderGugel  Can you add an example of this in use please?

**Comment by alexanderGugel** [comment]
```javascript
var domElement = new DOMElement(node);

// Option 1:

node.addComponent({
    onReceive: function(type, ev) {
        if (type === 'insertEl') {
            console.log('Inserted Element in DOM!!!', ev.el);
        }
    }
});

// Option 2:

// ... wait one frame ...

console.log(domElement.getEl()); // => Element   
```

Unfortunately there is no way around an asynchronous API.

**Comment by michaelobriena** [comment]
How does this work with the worker build?

**Comment by alexanderGugel** [comment]
Returns null. Objects that can't be processed through structured cloning will be stripped from _outCommands.

**Comment by DnMllr** [comment]
@alexanderGugel I am strongly against this. It is very important to me that our feature set and API are identical regardless of environment. There shouldn't be any punishment for working in a web worker.

The people who have needed references to the specific DOM element are mostly working on integrations. The way in which we support their desires cannot conflict with our own desire for platform independent rendering. For those who need access to the DOM, there is probably a way we can open up a space on the Render side of our platform to stick custom logic into the renderers. This would be a much better way to allow that functionality.

**Comment by alexanderGugel** [comment]
@DnMllr @zackbrown 

This has been a much requested feature (see Pivotal Tracker for original API proposal). I fully understand your concerns and agree for the most part with you from a **conceptual** point of view. On the other hand this has been a **much** requested feature, not only by people creating integrations, but also in general (has been asked couple of times on Slack).

> This would be a much better way to allow that functionality.

Adding this feature on the render size with some sort of register will result into a basically unusable API and users will be forced to entirely split there logic across threads.

Truth being told, it is not possible to not "punish" users in some way for using the Web worker. They won't ever have direct access to a DOM element. There is just no way around it. On the other hand, we can support the case without any problems when not running in a web worker, so we should do it IMO.

(just found a small bug, needs small fix, wait before merge :bug:)

**Comment by alexanderGugel** [comment]
Closing.

I'm still of the opinion that it has a significantly negative impact on the usability of the DOMElement if there is no ability to get the native Element.

**Comment by trusktr** [comment]
Maybe the native element never needs to be gotten? Perhaps `DOMElement.prototype.setContent` can accept HTMLElement references and add them to the `. famous-dom-element-content` element. People who use React, for example, could simply render their react component to a new HTMLElement, then pass that element into `setContent`. If they are in a web worker, then obviously they won't be able to have an HTMLElement to pass into setContent, and the APIs remain the same on both sides.

**Comment by trusktr** [comment]
Can a web worker even have a reference to a DOMElement to call setContent on it?

**Comment by alexanderGugel** [comment]
> If they are in a web worker, then they simply won't be able to have an HTMLElement to pass into setContent, thus the APIs remain the same on both sides.

I like that idea and it's in a way similar to the proposed solution for a related problem: Getting the reference to a native element via the Famous DOMElement.

> Can a web worker even have a reference to a DOMElement to call setContent on it?

No. It just wouldn't work in a Web Worker. I'm ok with this, but the overall decision has been made that we want the **exact** same environment, including the limitations associated with running in a Web Worker, even when running completely in the UI Thread.

So in short: No. And there is no useable solution for this problem.
